x-pack/plugins/security/public/components/breadcrumb.tsx

```diff
@@ -0,0 +1,134 @@
+
+export interface BreadcrumbProps extends EuiBreadcrumb {
+  text: string;
+}
+
+/**
+ * Component that automatically sets breadcrumbs and document title based on the render tree.
+ *
+ * @example
+ * // Breadcrumbs will be set to: "Users > Create"
+ * // Document title will be set to: "Create - Users"
+ *
+ * ```typescript
+ * <Breadcrumb text="Users">
+ *   <Table />
+ *   {showForm && (
+ *     <Breadcrumb text="Create">
+ *       <Form />
+ *     </Breadcrumb>
+ *   )}
+ * </Breadcrumb>
+ * ```
+ */
+export const Breadcrumb: FunctionComponent<BreadcrumbProps> = ({ children, ...breadcrumb }) => {
+  const context = useContext(BreadcrumbsContext);
+  const component = <InnerBreadcrumb breadcrumb={breadcrumb}>{children}</InnerBreadcrumb>;
+
+  if (context) {
+    return component;
+  }
+
+  return <Breadcrumbs>{component}</Breadcrumbs>;
+};
+
+export interface BreadcrumbsProps {
+  onChange?: BreadcrumbsChangeHandler;
+}
+
+export type BreadcrumbsChangeHandler = (breadcrumbs: BreadcrumbProps[]) => void;
+
+/**
+ * Component that can be used to define any side effects that should occur when breadcrumbs change.
+ *
+ * By default the breadcrumbs in application chrome are set and the document title is updated.
+ *
+ * @example
+ * ```typescript
+ * <Breadcrumbs onChange={(breadcrumbs) => setBreadcrumbs(breadcrumbs)}>
+ *   <Breadcrumb text="Users" />
+ * </Breadcrumbs>
+ * ```
+ */
+export const Breadcrumbs: FunctionComponent<BreadcrumbsProps> = ({ children, onChange }) => {
+  const { services } = useKibana();
+  const breadcrumbsRef = useRef<BreadcrumbProps[]>([]);
+
+  const handleChange = (breadcrumbs: BreadcrumbProps[]) => {
+    if (onChange) {
+      onChange(breadcrumbs);
+    } else if (services.chrome) {
+      services.chrome.setBreadcrumbs(breadcrumbs);
+      services.chrome.docTitle.change(getDocTitle(breadcrumbs));
+    }
+  };
+
+  return (
+    <BreadcrumbsContext.Provider
+      value={{
+        parents: [],
+        onMount: (breadcrumbs) => {
+          if (breadcrumbs.length > breadcrumbsRef.current.length) {
+            breadcrumbsRef.current = breadcrumbs;
+            handleChange(breadcrumbs);
+          }
+        },
+        onUnmount: (breadcrumbs) => {
+          if (breadcrumbs.length < breadcrumbsRef.current.length) {
+            breadcrumbsRef.current = breadcrumbs;
+            handleChange(breadcrumbs);
+          }
+        },
+      }}
+    >
+      {children}
+    </BreadcrumbsContext.Provider>
+  );
+};
+
+export interface InnerBreadcrumbProps {
+  breadcrumb: BreadcrumbProps;
+}
+
+export const InnerBreadcrumb: FunctionComponent<InnerBreadcrumbProps> = ({
+  breadcrumb,
+  children,
+}) => {
+  const { parents, onMount, onUnmount } = useContext(BreadcrumbsContext)!;
+  const nextParents = [...parents, breadcrumb];
+
+  useEffect(() => {
+    onMount(nextParents);
+    return () => onUnmount(parents);
+  }, [breadcrumb.text, breadcrumb.href]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
In this particular case, I want to change the breadcrumbs in Kibana when:

* A new Breadcrumb component is mounted
* An existing Breadcrumb component received a different `text` prop
* An existing Breadcrumb component received a different `href` prop
* An existing Breadcrumb component is unmounted

The `nextParents` array is created on every render. If I blindly follow the exhaustive-deps eslint rule I would cause the `services.chrome.breadcrumbs.setBreadcrumbs` method to be called with every render of every Breadcrumb component. 

The rule is generally doing more harm than good as it assumes that every non-primitive value (callbacks and objects) will be memoized. In reality that rarely happens. 

We can follow the eslint rule but then there should almost be another rule to ensure that non-primitives are memoized. We'd also need to consider that memoization itself is "work" that the CPU needs to carry out and takes up memory, which is why I went with this approach. 

It's definitely not perfect though. For instance, if a Breadcrumb intentionally gets a new click handler that does something different than the previous one, the current implementation would not update the Kibana header since `breadcrumbs.onClick` is not in the dependency array. 

Again, that was intentional since most devs use the `reactRouterNavigate` method provided by core that creates a new function with every render. If we were to use exhaustive-deps those would need to be memoized before being passed into the Breadcrumb to avoid unnecessary re-renders. 

It's a complicated issue with different tradeoffs. Happy to re-assess and discuss further.

Reply:
Thanks for taking the time to explain - this makes sense to me now. I'm absolutely fine to leave this the way you have it. The tradeoffs that you took here seem reasonable to me.